universum.cabal

```diff
@@ -105,7 +105,7 @@ library
                      , stm
                      -- Make sure that "toString-toText-rewritting" note
                      -- is still valid when bumping this constraint.
-                     , text >= 1.0.0.0 && <= 1.2.5.0
+                     , text >= 1.0.0.0 && <= 2.0.1
```

Comment:
According to the [PVP](https://pvp.haskell.org/):

![image](https://user-images.githubusercontent.com/718166/220048189-7ecc455c-ff6c-4476-a98b-ae9fa5d33dd3.png)

Is `universum` insensitive to **additions** to `text`'s API? If so (I think it is :thinking:), then perhaps it would be better to set the upper bound to `< 2.1`

Reply:
We have some rewrite rules that depend on the implementation details of some functions in `text`: https://github.com/serokell/universum/blob/0ee8f07e7200fd35fe6be2abbe50e96f7f5acc6e/src/Universum/String/Conversion.hs#L182

The goal is to make sure that `toString . toText` will be replaced with `id`, because usually it's what you want and it can give a good performance improvement.
I think it's possible to change the implementation of `Text.pack` or `Text.unpack` without affecting any types or semantics of any function. So it can be done even without changing A.B version components. So we have a very strict upper bound because we depend on implementation details that normally we shouldn't depend on.

Btw, in PVP they have:
> 1. Breaking change. If any entity was removed, or the types of any entities or the definitions of datatypes or classes were changed, or orphan instances were added or any instances were removed,
> 2. Non-breaking change. Otherwise, if only new bindings, types, classes, non-orphan instances or modules (but see below) were added to the interface
> 3. Other changes. Otherwise, e.g. if change consist only of corrected documentation, non-visible change to allow different dependency range etc.

If I change `Text.pack` to drop invalid `Char`s instead of replacing them, does it count as (1) or (2) or (3)? Naturally, I think it's a breaking change and should count as (1). But they say:
> the types of any entities or the definitions of datatypes or classes were changed

In my example the types do not change and the definitions of datatypes and classes do not change either. Only the definition of a function changes.

Anyway, I think there can be other changes that do not affect semantics of any exported function, but make the rule `forall s. TF.unstreamList (TF.map T.safe (TF.streamList s)) = s` useless.

Reply:
Thank you for the comprehensive explanation :)